File: src/main/java/no/nav/bidrag/beregn/bidragsevne/beregning/BidragsevneBeregningImpl.java
package no.nav.bidrag.beregn.bidragsevne.beregning;

import static java.util.Collections.singletonList;
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.toList;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import no.nav.bidrag.beregn.bidragsevne.bo.GrunnlagBeregning;
import no.nav.bidrag.beregn.bidragsevne.bo.Inntekt;
import no.nav.bidrag.beregn.bidragsevne.bo.ResultatBeregning;
import no.nav.bidrag.beregn.felles.SjablonUtil;
import no.nav.bidrag.beregn.felles.bo.Periode;
import no.nav.bidrag.beregn.felles.bo.SjablonNokkel;
import no.nav.bidrag.beregn.felles.bo.SjablonPeriode;
import no.nav.bidrag.beregn.felles.bo.SjablonPeriodeNavnVerdi;
import no.nav.bidrag.beregn.felles.bo.TrinnvisSkattesats;
import no.nav.bidrag.beregn.felles.enums.BostatusKode;
import no.nav.bidrag.beregn.felles.enums.SaerfradragKode;
import no.nav.bidrag.beregn.felles.enums.SjablonInnholdNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonNokkelNavn;
import no.nav.bidrag.beregn.felles.enums.SjablonTallNavn;

public class BidragsevneBeregningImpl implements BidragsevneBeregning {

  @Override
  public ResultatBeregning beregn(GrunnlagBeregning grunnlagBeregning) {

    // Henter sjablonverdier
    var sjablonNavnVerdiMap = hentSjablonVerdier(grunnlagBeregning.getSjablonListe(), grunnlagBeregning.getBostatus().getKode(),
        grunnlagBeregning.getSkatteklasse().getSkatteklasse());

    // Beregn minstefradrag
    var minstefradrag = beregnMinstefradrag(grunnlagBeregning,
        sjablonNavnVerdiMap.get(SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP.getNavn()),
        sjablonNavnVerdiMap.get(SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT.getNavn()));

    // Legger sammen inntektene
    var inntekt = grunnlagBeregning.getInntektListe().stream()
        .map(Inntekt::getBelop)
        .reduce(BigDecimal.ZERO, BigDecimal::add);

    // finner 25% av inntekt og omregner til månedlig beløp
    var tjuefemProsentInntekt = inntekt
        .divide(BigDecimal.valueOf(4), new MathContext(10, RoundingMode.HALF_UP))
        .divide(BigDecimal.valueOf(12), new MathContext(10, RoundingMode.HALF_UP));

    tjuefemProsentInntekt = tjuefemProsentInntekt.setScale(0, RoundingMode.HALF_UP);

    // finner personfradragklasse ut fra angitt skatteklasse
    var personfradrag = BigDecimal.ZERO;
    if (grunnlagBeregning.getSkatteklasse().getSkatteklasse() == 1) {
      personfradrag = sjablonNavnVerdiMap.get(SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELOP.getNavn());
    } else {
      personfradrag = sjablonNavnVerdiMap.get(SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELOP.getNavn());
    }

    var inntektMinusFradrag = inntekt.subtract(minstefradrag).subtract(personfradrag);

    // Trekker fra skatt
    var forelopigBidragsevne = inntekt.subtract(inntektMinusFradrag.multiply(
        sjablonNavnVerdiMap.get(SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT.getNavn()).divide(BigDecimal.valueOf(100),
            new MathContext(10, RoundingMode.HALF_UP))));

    // Trekker fra trygdeavgift
    forelopigBidragsevne = (forelopigBidragsevne.subtract((inntekt.multiply(
        sjablonNavnVerdiMap.get(SjablonTallNavn.TRYGDEAVGIFT_PROSENT.getNavn()).divide(BigDecimal.valueOf(100),
            new MathContext(10, RoundingMode.HALF_UP))))));

    // Trekker fra trinnvis skatt
    forelopigBidragsevne = forelopigBidragsevne.subtract(beregnSkattetrinnBelop(grunnlagBeregning));

    // Trekker fra boutgifter og midler til eget underhold
    forelopigBidragsevne = forelopigBidragsevne.subtract(
        sjablonNavnVerdiMap.get(SjablonInnholdNavn.BOUTGIFT_BELOP.getNavn()).multiply(BigDecimal.valueOf(12)));

    forelopigBidragsevne = forelopigBidragsevne.subtract(
        sjablonNavnVerdiMap.get(SjablonInnholdNavn.UNDERHOLD_BELOP.getNavn()).multiply(BigDecimal.valueOf(12)));

    // Trekker fra midler til underhold egne barn i egen husstand
    forelopigBidragsevne = forelopigBidragsevne.subtract(
        sjablonNavnVerdiMap.get(SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELOP.getNavn())
            .multiply(BigDecimal.valueOf(grunnlagBeregning.getBarnIHusstand().getAntallBarn()))
            .multiply(BigDecimal.valueOf(12)));

    // Sjekker om og kalkulerer eventuell fordel særfradrag
    if (grunnlagBeregning.getSaerfradrag().getKode().equals(SaerfradragKode.HELT)) {
      forelopigBidragsevne = forelopigBidragsevne.add(sjablonNavnVerdiMap.get(SjablonTallNavn.FORDEL_SAERFRADRAG_BELOP.getNavn()));
    } else if (grunnlagBeregning.getSaerfradrag().getKode().equals(SaerfradragKode.HALVT)) {
      forelopigBidragsevne = forelopigBidragsevne.add(
          sjablonNavnVerdiMap.get(SjablonTallNavn.FORDEL_SAERFRADRAG_BELOP.getNavn()).divide(BigDecimal.valueOf(2),
              new MathContext(10, RoundingMode.HALF_UP)));
    }

    // Legger til fordel skatteklasse2
    if (grunnlagBeregning.getSkatteklasse().getSkatteklasse() == 2) {
      forelopigBidragsevne = forelopigBidragsevne.add(sjablonNavnVerdiMap.get(SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.getNavn()));
    }

    // Finner månedlig beløp for bidragsevne
    var maanedligBidragsevne = forelopigBidragsevne
        .divide(BigDecimal.valueOf(12), new MathContext(10, RoundingMode.HALF_UP))
        .setScale(0, RoundingMode.HALF_UP);

    if (maanedligBidragsevne.compareTo(BigDecimal.ZERO) < 0) {
      maanedligBidragsevne = BigDecimal.ZERO;
    }

    return new ResultatBeregning(maanedligBidragsevne, tjuefemProsentInntekt, byggSjablonResultatListe(sjablonNavnVerdiMap,
        grunnlagBeregning.getSjablonListe()));
  }

  private BigDecimal beregnMinstefradrag(GrunnlagBeregning grunnlagBeregning, BigDecimal minstefradragInntektSjablonBelop,
      BigDecimal minstefradragInntektSjablonProsent) {

    // Legger sammen inntektene
    var inntekt = grunnlagBeregning.getInntektListe().stream()
        .map(Inntekt::getBelop)
        .reduce(BigDecimal.ZERO, BigDecimal::add);

    var minstefradrag = inntekt.multiply(minstefradragInntektSjablonProsent.divide(BigDecimal.valueOf(100),
        new MathContext(2, RoundingMode.HALF_UP)));

    if (minstefradrag.compareTo(minstefradragInntektSjablonBelop) > 0) {
      minstefradrag = minstefradragInntektSjablonBelop;
    }

    minstefradrag = minstefradrag.setScale(0, RoundingMode.HALF_UP);

    return minstefradrag;
  }

  private BigDecimal beregnSkattetrinnBelop(GrunnlagBeregning grunnlagBeregning) {

    // Legger sammen inntektene
    var inntekt = grunnlagBeregning.getInntektListe().stream()
        .map(Inntekt::getBelop)
        .reduce(BigDecimal.ZERO, BigDecimal::add);

    var sortertTrinnvisSkattesatsListe = SjablonUtil.hentTrinnvisSkattesats(
        grunnlagBeregning.getSjablonListe().stream()
            .map(SjablonPeriode::getSjablon)
            .collect(toList()),
        SjablonNavn.TRINNVIS_SKATTESATS);

    BigDecimal samletSkattetrinnBelop = BigDecimal.ZERO;
    var indeks = 1;

    // Beregn skattetrinnbeløp
    while (indeks < sortertTrinnvisSkattesatsListe.size()) {
      if (inntekt.compareTo(sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense()) > 0) {
        if (inntekt.compareTo(sortertTrinnvisSkattesatsListe.get(indeks).getInntektGrense()) < 0) {
          samletSkattetrinnBelop = samletSkattetrinnBelop.add(inntekt.subtract(
              sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense())
              .multiply(sortertTrinnvisSkattesatsListe.get(indeks - 1).getSats())
              .divide(BigDecimal.valueOf(100), new MathContext(10, RoundingMode.HALF_UP)));
        } else {
          samletSkattetrinnBelop = samletSkattetrinnBelop.add(
              sortertTrinnvisSkattesatsListe.get(indeks).getInntektGrense()
                  .subtract(sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense())
                  .multiply(sortertTrinnvisSkattesatsListe.get(indeks - 1).getSats()
                      .divide(BigDecimal.valueOf(100), new MathContext(10, RoundingMode.HALF_UP))));
        }
      }
      indeks = indeks + 1;
    }

    if (inntekt.compareTo(sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense()) > 0) {
      samletSkattetrinnBelop = samletSkattetrinnBelop.add(
          (inntekt.subtract(sortertTrinnvisSkattesatsListe.get(indeks - 1).getInntektGrense())
              .multiply(sortertTrinnvisSkattesatsListe.get(indeks - 1).getSats())
              .divide(BigDecimal.valueOf(100),
                  new MathContext(1, RoundingMode.HALF_UP))));
    }

    samletSkattetrinnBelop = samletSkattetrinnBelop.setScale(0, RoundingMode.HALF_UP);

    return samletSkattetrinnBelop;
  }

  // Henter sjablonverdier
  private Map<String, BigDecimal> hentSjablonVerdier(List<SjablonPeriode> sjablonPeriodeListe, BostatusKode bostatusKode, int skatteklasse) {

    var sjablonListe = sjablonPeriodeListe.stream()
        .map(SjablonPeriode::getSjablon)
        .collect(toList());

    var sjablonNavnVerdiMap = new HashMap<String, BigDecimal>();

    // Sjablontall
    if (skatteklasse == 1) {
      sjablonNavnVerdiMap.put(SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELOP.getNavn(),
          SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELOP));
    } else {
      sjablonNavnVerdiMap.put(SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELOP.getNavn(),
          SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELOP));
      sjablonNavnVerdiMap.put(SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP.getNavn(),
          SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELOP));
    }
    sjablonNavnVerdiMap.put(SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT.getNavn(),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT));
    sjablonNavnVerdiMap.put(SjablonTallNavn.TRYGDEAVGIFT_PROSENT.getNavn(),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.TRYGDEAVGIFT_PROSENT));
    sjablonNavnVerdiMap.put(SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELOP.getNavn(),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELOP));
    sjablonNavnVerdiMap.put(SjablonTallNavn.FORDEL_SAERFRADRAG_BELOP.getNavn(),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.FORDEL_SAERFRADRAG_BELOP));
    sjablonNavnVerdiMap.put(SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT.getNavn(),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT));
    sjablonNavnVerdiMap.put(SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP.getNavn(),
        SjablonUtil.hentSjablonverdi(sjablonListe, SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELOP));

    // Bidragsevne
    var sjablonNokkelVerdi = bostatusKode.equals(BostatusKode.ALENE) ? "EN" : "GS";
    sjablonNavnVerdiMap.put(SjablonInnholdNavn.BOUTGIFT_BELOP.getNavn(), SjablonUtil.hentSjablonverdi(sjablonListe, SjablonNavn.BIDRAGSEVNE,
        singletonList(new SjablonNokkel(SjablonNokkelNavn.BOSTATUS.getNavn(), sjablonNokkelVerdi)), SjablonInnholdNavn.BOUTGIFT_BELOP));
    sjablonNavnVerdiMap.put(SjablonInnholdNavn.UNDERHOLD_BELOP.getNavn(), SjablonUtil.hentSjablonverdi(sjablonListe, SjablonNavn.BIDRAGSEVNE,
        singletonList(new SjablonNokkel(SjablonNokkelNavn.BOSTATUS.getNavn(), sjablonNokkelVerdi)), SjablonInnholdNavn.UNDERHOLD_BELOP));

    // TrinnvisSkattesats
    var trinnvisSkattesatsListe = SjablonUtil.hentTrinnvisSkattesats(sjablonListe, SjablonNavn.TRINNVIS_SKATTESATS);
    var indeks = 1;
    for (TrinnvisSkattesats trinnvisSkattesats : trinnvisSkattesatsListe) {
      sjablonNavnVerdiMap.put(SjablonNavn.TRINNVIS_SKATTESATS.getNavn() + "InntektGrense" + indeks, trinnvisSkattesats.getInntektGrense());
      sjablonNavnVerdiMap.put(SjablonNavn.TRINNVIS_SKATTESATS.getNavn() + "Sats" + indeks, trinnvisSkattesats.getSats());
      indeks++;
    }

    return sjablonNavnVerdiMap;
  }

  // Mapper ut sjablonverdier til ResultatBeregning (dette for å sikre at kun sjabloner som faktisk er brukt legges ut i grunnlaget for beregning)
  private List<SjablonPeriodeNavnVerdi> byggSjablonResultatListe(Map<String, BigDecimal> sjablonNavnVerdiMap,
      List<SjablonPeriode> sjablonPeriodeListe) {
    var sjablonPeriodeNavnVerdiListe = new ArrayList<SjablonPeriodeNavnVerdi>();
    sjablonNavnVerdiMap.forEach((sjablonNavn, sjablonVerdi) ->
        sjablonPeriodeNavnVerdiListe.add(new SjablonPeriodeNavnVerdi(hentPeriode(sjablonPeriodeListe, sjablonNavn), sjablonNavn, sjablonVerdi)));
    return sjablonPeriodeNavnVerdiListe.stream().sorted(comparing(SjablonPeriodeNavnVerdi::getNavn)).collect(toList());
  }

  private Periode hentPeriode(List<SjablonPeriode> sjablonPeriodeListe, String sjablonNavn) {
    return sjablonPeriodeListe.stream()
        .filter(sjablonPeriode -> sjablonPeriode.getSjablon().getNavn().equals(sjablonNavn))
        .map(SjablonPeriode::getPeriode)
        .findFirst()
        .orElse(new Periode(LocalDate.MIN, LocalDate.MAX));
  }
}

File: src/main/kotlin/no/nav/bidrag/beregn/bidragsevne/beregning/BidragsevneBeregningImpl.kt
package no.nav.bidrag.beregn.bidragsevne.beregning

import no.nav.bidrag.beregn.bidragsevne.bo.GrunnlagBeregning
import no.nav.bidrag.beregn.bidragsevne.bo.ResultatBeregning
import no.nav.bidrag.beregn.felles.FellesBeregning
import no.nav.bidrag.beregn.felles.bo.SjablonNokkel
import no.nav.bidrag.beregn.felles.bo.SjablonPeriode
import no.nav.bidrag.beregn.felles.util.SjablonUtil
import no.nav.bidrag.beregn.felles.util.SjablonUtil.hentSjablonverdi
import no.nav.bidrag.domene.enums.beregning.Særfradragskode
import no.nav.bidrag.domene.enums.person.Bostatuskode
import no.nav.bidrag.domene.enums.sjablon.SjablonInnholdNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonNøkkelNavn
import no.nav.bidrag.domene.enums.sjablon.SjablonTallNavn
import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode

open class BidragsevneBeregningImpl : FellesBeregning(), BidragsevneBeregning {
    override fun beregn(grunnlag: GrunnlagBeregning): ResultatBeregning {
        // Henter sjablonverdier
        val sjablonNavnVerdiMap =
            hentSjablonVerdier(
                sjablonPeriodeListe = grunnlag.sjablonListe,
                bostatusKode = grunnlag.bostatus.kode,
                skatteklasse = grunnlag.skatteklasse.skatteklasse,
            )

        // Beregner minstefradrag
        val minstefradrag =
            beregnMinstefradrag(
                grunnlag = grunnlag,
                minstefradragInntektSjablonBelop = sjablonNavnVerdiMap[SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELØP.navn] ?: BigDecimal.ZERO,
                minstefradragInntektSjablonProsent = sjablonNavnVerdiMap[SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT.navn] ?: BigDecimal.ZERO,
            )

        // Legger sammen inntektene
        val inntekt = grunnlag.inntektListe.fold(BigDecimal.ZERO) { total, inntekt -> total + inntekt.belop }

        // finner 25% av inntekt og omregner til månedlig beløp
        val tjuefemProsentInntekt =
            inntekt
                .divide(BigDecimal.valueOf(4), MathContext(10, RoundingMode.HALF_UP))
                .divide(BigDecimal.valueOf(12), MathContext(10, RoundingMode.HALF_UP))
                .setScale(0, RoundingMode.HALF_UP)

        // finner personfradragklasse ut fra angitt skatteklasse
        val personfradrag =
            if (grunnlag.skatteklasse.skatteklasse == 1) {
                sjablonNavnVerdiMap[SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELØP.navn]
            } else {
                sjablonNavnVerdiMap[SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELØP.navn]
            }

        val inntektMinusFradrag = inntekt - minstefradrag - (personfradrag ?: BigDecimal.ZERO)

        // Trekker fra skatt
        var forelopigBidragsevne =
            inntekt - (
                inntektMinusFradrag *
                    (sjablonNavnVerdiMap[SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT.navn] ?: BigDecimal.ZERO)
                        .divide(BigDecimal.valueOf(100), MathContext(10, RoundingMode.HALF_UP))
                )

        // Trekker fra trygdeavgift
        forelopigBidragsevne -= inntekt *
            (sjablonNavnVerdiMap[SjablonTallNavn.TRYGDEAVGIFT_PROSENT.navn] ?: BigDecimal.ZERO)
                .divide(BigDecimal.valueOf(100), MathContext(10, RoundingMode.HALF_UP))

        // Trekker fra trinnvis skatt
        forelopigBidragsevne -= beregnSkattetrinnBelop(grunnlag)

        // Trekker fra boutgifter og midler til eget underhold
        forelopigBidragsevne -= (sjablonNavnVerdiMap[SjablonInnholdNavn.BOUTGIFT_BELØP.navn] ?: BigDecimal.ZERO) * BigDecimal.valueOf(12)
        forelopigBidragsevne -= (sjablonNavnVerdiMap[SjablonInnholdNavn.UNDERHOLD_BELØP.navn] ?: BigDecimal.ZERO) * BigDecimal.valueOf(12)

        // Trekker fra midler til underhold egne barn i egen husstand
        forelopigBidragsevne -= (
            sjablonNavnVerdiMap[SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELØP.navn]
                ?: BigDecimal.ZERO
            ) * BigDecimal.valueOf(grunnlag.barnIHusstand.antallBarn) * BigDecimal.valueOf(12)

        // Sjekker om og kalkulerer eventuell fordel særfradrag
        forelopigBidragsevne =
            when (grunnlag.saerfradrag.kode) {
                Særfradragskode.HELT -> forelopigBidragsevne + (sjablonNavnVerdiMap[SjablonTallNavn.FORDEL_SÆRFRADRAG_BELØP.navn] ?: BigDecimal.ZERO)
                Særfradragskode.HALVT ->
                    forelopigBidragsevne +
                        (sjablonNavnVerdiMap[SjablonTallNavn.FORDEL_SÆRFRADRAG_BELØP.navn] ?: BigDecimal.ZERO)
                            .divide(BigDecimal.valueOf(2), MathContext(10, RoundingMode.HALF_UP))

                else -> forelopigBidragsevne
            }

        // Legger til fordel skatteklasse2
        if (grunnlag.skatteklasse.skatteklasse == 2) {
            forelopigBidragsevne += sjablonNavnVerdiMap[SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELØP.navn] ?: BigDecimal.ZERO
        }

        // Finner månedlig beløp for bidragsevne
        val maanedligBidragsevne =
            forelopigBidragsevne
                .divide(BigDecimal.valueOf(12), MathContext(10, RoundingMode.HALF_UP))
                .max(BigDecimal.ZERO)
                .setScale(0, RoundingMode.HALF_UP)

        return ResultatBeregning(
            belop = maanedligBidragsevne,
            inntekt25Prosent = tjuefemProsentInntekt,
            sjablonListe = byggSjablonResultatListe(sjablonNavnVerdiMap, grunnlag.sjablonListe),
        )
    }

    private fun beregnMinstefradrag(
        grunnlag: GrunnlagBeregning,
        minstefradragInntektSjablonBelop: BigDecimal,
        minstefradragInntektSjablonProsent: BigDecimal,
    ): BigDecimal {
        // Legger sammen inntektene
        val inntekt = grunnlag.inntektListe.fold(BigDecimal.ZERO) { total, inntekt -> total + inntekt.belop }

        return (inntekt * minstefradragInntektSjablonProsent.divide(BigDecimal.valueOf(100), MathContext(2, RoundingMode.HALF_UP)))
            .min(minstefradragInntektSjablonBelop)
            .setScale(0, RoundingMode.HALF_UP)
    }

    private fun beregnSkattetrinnBelop(grunnlag: GrunnlagBeregning): BigDecimal {
        // Legger sammen inntektene
        val inntekt = grunnlag.inntektListe.fold(BigDecimal.ZERO) { total, inntekt -> total + inntekt.belop }

        val sortertTrinnvisSkattesatsListe =
            SjablonUtil.hentTrinnvisSkattesats(
                sjablonListe = grunnlag.sjablonListe.map { it.sjablon },
                sjablonNavn = SjablonNavn.TRINNVIS_SKATTESATS,
            )

        var samletSkattetrinnBelop = BigDecimal.ZERO
        var indeks = 1

        // Beregn skattetrinnbeløp
        while (indeks < sortertTrinnvisSkattesatsListe.size) {
            val denneGrense = sortertTrinnvisSkattesatsListe[indeks - 1].inntektGrense
            val nesteGrense = sortertTrinnvisSkattesatsListe[indeks].inntektGrense

            if (inntekt > denneGrense) {
                val taxableIncome = minOf(inntekt, nesteGrense) - denneGrense
                val taxRate = sortertTrinnvisSkattesatsListe[indeks - 1].sats

                samletSkattetrinnBelop += (taxableIncome * taxRate).divide(BigDecimal.valueOf(100), MathContext(10, RoundingMode.HALF_UP))
            }
            indeks++
        }

        if (inntekt > sortertTrinnvisSkattesatsListe[indeks - 1].inntektGrense) {
            val taxableIncome = inntekt - sortertTrinnvisSkattesatsListe[indeks - 1].inntektGrense
            val taxRate = sortertTrinnvisSkattesatsListe[indeks - 1].sats

            samletSkattetrinnBelop += (taxableIncome * taxRate / BigDecimal.valueOf(100)).setScale(1, RoundingMode.HALF_UP)
        }

        return samletSkattetrinnBelop.setScale(0, RoundingMode.HALF_UP)
    }

    // Henter sjablonverdier
    private fun hentSjablonVerdier(
        sjablonPeriodeListe: List<SjablonPeriode>,
        bostatusKode: Bostatuskode,
        skatteklasse: Int,
    ): Map<String, BigDecimal> {
        val sjablonNavnVerdiMap = HashMap<String, BigDecimal>()
        val sjablonListe = sjablonPeriodeListe.map { it.sjablon }

        // Sjablontall
        if (skatteklasse == 1) {
            sjablonNavnVerdiMap[SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELØP.navn] =
                hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.PERSONFRADRAG_KLASSE1_BELØP)
        } else {
            sjablonNavnVerdiMap[SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELØP.navn] =
                hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.PERSONFRADRAG_KLASSE2_BELØP)
            sjablonNavnVerdiMap[SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELØP.navn] =
                hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.FORDEL_SKATTEKLASSE2_BELØP)
        }

        sjablonNavnVerdiMap[SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT.navn] =
            hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.SKATTESATS_ALMINNELIG_INNTEKT_PROSENT)
        sjablonNavnVerdiMap[SjablonTallNavn.TRYGDEAVGIFT_PROSENT.navn] =
            hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.TRYGDEAVGIFT_PROSENT)
        sjablonNavnVerdiMap[SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELØP.navn] =
            hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.UNDERHOLD_EGNE_BARN_I_HUSSTAND_BELØP)
        sjablonNavnVerdiMap[SjablonTallNavn.FORDEL_SÆRFRADRAG_BELØP.navn] =
            hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.FORDEL_SÆRFRADRAG_BELØP)
        sjablonNavnVerdiMap[SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT.navn] =
            hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.MINSTEFRADRAG_INNTEKT_PROSENT)
        sjablonNavnVerdiMap[SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELØP.navn] =
            hentSjablonverdi(sjablonListe = sjablonListe, sjablonTallNavn = SjablonTallNavn.MINSTEFRADRAG_INNTEKT_BELØP)

        // Bidragsevne
        val sjablonNokkelVerdi = if (bostatusKode == Bostatuskode.ALENE) "EN" else "GS"
        sjablonNavnVerdiMap[SjablonInnholdNavn.BOUTGIFT_BELØP.navn] =
            hentSjablonverdi(
                sjablonListe = sjablonListe,
                sjablonNavn = SjablonNavn.BIDRAGSEVNE,
                sjablonNokkelListe = listOf(SjablonNokkel(navn = SjablonNøkkelNavn.BOSTATUS.navn, verdi = sjablonNokkelVerdi)),
                sjablonInnholdNavn = SjablonInnholdNavn.BOUTGIFT_BELØP,
            )
        sjablonNavnVerdiMap[SjablonInnholdNavn.UNDERHOLD_BELØP.navn] =
            hentSjablonverdi(
                sjablonListe = sjablonListe,
                sjablonNavn = SjablonNavn.BIDRAGSEVNE,
                sjablonNokkelListe = listOf(SjablonNokkel(SjablonNøkkelNavn.BOSTATUS.navn, sjablonNokkelVerdi)),
                sjablonInnholdNavn = SjablonInnholdNavn.UNDERHOLD_BELØP,
            )

        // TrinnvisSkattesats
        val trinnvisSkattesatsListe =
            SjablonUtil.hentTrinnvisSkattesats(
                sjablonListe = sjablonListe,
                sjablonNavn = SjablonNavn.TRINNVIS_SKATTESATS,
            )
        var indeks = 1
        trinnvisSkattesatsListe.forEach {
            sjablonNavnVerdiMap[SjablonNavn.TRINNVIS_SKATTESATS.navn + "InntektGrense" + indeks] = it.inntektGrense
            sjablonNavnVerdiMap[SjablonNavn.TRINNVIS_SKATTESATS.navn + "Sats" + indeks] = it.sats
            indeks++
        }

        return sjablonNavnVerdiMap
    }
}
